// vite-plugin-vue-meta-layouts@0.6.1/dist/index.cjs
'use strict';

const path = require('path');
const localPkg = require('local-pkg');

function normalizePath(path$1) {
  path$1 = path$1.startsWith("/") ? path$1 : "/" + path$1;
  return path.posix.normalize(path$1);
}
async function isVite2() {
  const info = await localPkg.getPackageInfo("vite");
  if (info) {
    return /^.?2/.test(info.version);
  }
  return false;
}

function createVirtualModuleID(name) {
  const virtualModuleId = `virtual:${name}`;
  const resolvedVirtualModuleId = "\0" + virtualModuleId;
  return {
    virtualModuleId,
    resolvedVirtualModuleId
  };
}
async function createVirtualGlob(target, isSync, excludes) {
  const excludeGlobs = excludes.map((exclude) => `"!${target}/${exclude}"`).join(",");
  const g = excludeGlobs === "" ? `"${target}/**/*.vue"` : `["${target}/**/*.vue", ${excludeGlobs}]`;
  if (await isVite2()) {
    return isSync ? `import.meta.globEager(${g})` : `import.meta.glob(${g})`;
  }
  return `import.meta.glob(${g}, { eager: ${isSync} })`;
}
async function createVirtualModuleCode(options) {
  const {
    target,
    defaultLayout,
    importMode,
    skipTopLevelRouteLayout,
    excludes,
    metaName
  } = options;
  const normalizedTarget = normalizePath(target);
  const isSync = importMode === "sync";
  const skipCode = `// unplugin-vue-router adds a top-level route to the routing group, which we should skip. (ref \u2192 https://github.com/JohnCampionJr/vite-plugin-vue-layouts/issues/134)
  const skipLayout = !route.component && route.children?.find(r => (r.path === '' || r.path === '/') && r.meta?.isLayout)  

  if (skipLayout) {
    return route
  }`;
  return `
export function createGetRoutes(router, withLayout = false) {
	const routes = router.getRoutes()
	if (withLayout) {
		return routes
	}
	return () => routes.filter(route => !route.meta.isLayout)
}

export function setupLayouts(routes) {
	const layouts = {}

	const modules = ${await createVirtualGlob(
    normalizedTarget,
    isSync,
    excludes
  )}

	Object.entries(modules).forEach(([name, module]) => {
		let key = name.replace("${normalizedTarget}/", '').replace('.vue', '')
		layouts[key] = ${isSync ? "module.default" : "module"}
	})
	
  function deepSetupLayout(routes, top = true) {
    return routes.map(route => {
      if (route.children?.length > 0) {
        route.children = deepSetupLayout(route.children, false)
      }
      
      if (top) {
        ${skipTopLevelRouteLayout ? skipCode : ""}
        if (route.meta?.${metaName} !== false) {
          return {
            path: route.path,
            component: layouts[route.meta?.${metaName} || '${defaultLayout}'],
            // ref \u2192 https://github.com/JohnCampionJr/vite-plugin-vue-layouts/pull/97
            children: route.path === '/' ? [route] : [{...route, path: ''}],
            meta: {
              isLayout: true
            }
          }
        }
      }

      if (route.meta?.${metaName}) {
        return { 
          path: route.path,
          component: layouts[route.meta?.${metaName}],
          children: [ {...route, path: ''} ],
          meta: {
            isLayout: true
          }
        }
      }

      return route
    })
  }

	return deepSetupLayout(routes)
}`;
}

function MetaLayouts(options = {}) {
  const {
    target = "src/layouts",
    defaultLayout = "default",
    importMode = process.env.VITE_SSG ? "sync" : "async",
    skipTopLevelRouteLayout = false,
    excludes = [],
    metaName = "layout"
  } = options;
  const { virtualModuleId, resolvedVirtualModuleId } = createVirtualModuleID(
    "meta-layouts"
  );
  return {
    name: "vite-plugin-vue-meta-layouts",
    resolveId(id) {
      if (id === virtualModuleId) {
        return resolvedVirtualModuleId;
      }
    },
    load(id) {
      if (id === resolvedVirtualModuleId) {
        return createVirtualModuleCode({
          target,
          importMode,
          defaultLayout,
          skipTopLevelRouteLayout,
          excludes,
          metaName
        });
      }
    }
  };
}

module.exports = MetaLayouts;
